### DISCUSSION TITLE
Support zlib engine like openssl engine

### DISCUSSION THREAD

**Discussion by Ocean-Lua** [zlib-ng/zlib-ng#1301] [Ideas]
By providing an engine framework to support third-party hardware acceleration, the application can obtain hardware acceleration without modifying the code.

**Comment by mtl1979** [comment]
Once the zlib-ng API and the internals are stabilized, we can add frameworks and wrappers more easily... Some cases it might make sense to separate them as own repository so the build time on main repository doesn't increase too much and we don't require too many dependencies that we need to implicitly download...

**Reply by dingwei-2017** [reply]
hi，mtl1979.  we also need to support third-party expansion for zlib-ng, just like openssl. From your comment above, i guess you also have the will to support this ability.  If so, we can have a discussion about how to enable this feature in detail. We will be exceedingly honoured to participate in the development of this feature.

**Reply by mtl1979** [reply]
One of the key features of zlib-ng is to be more modern replacement of zlib. This also means we should support more modern features... zlib-ng already has abstracted a lot of necessary code paths, but we will need to add optional hooks to some of the functions that for example detect CPU features, and initialize `functable` depending on what CPU features are available...

What I know about hook functions is that they essentially return boolean that tells if the calling function should continue execution or return prematurely due to one of the hook functions having taken care of the feature. If the hook function doesn't take care of the feature, it will relay the parameters to the next hook in the chain until there is no hooks left. Last hook in the chain signals that the calling function should continue the execution to the "internal" code path. For some hook functions, like checking the CPU features, the return value is unnecessary and can be omitted.

Caveat is that having too many hooks might slow down the initialization of `functable`, but in most cases there will not be enough hook functions to delay the initialization for more than few milliseconds. It's kind of one time cost per feature.

**Reply by nmoinvaz** [reply]
Can anybody provide more exact detail on what you are looking for?

**Reply by mtl1979** [reply]
It's a lot like what inflateBack or the DFLTCC hooks do... Basically override some parts of zlib-ng to add better implementation... Then to add for example support for more operating systems or I/O stream types, the code on main repository doesn't need to be recompiled at all...
